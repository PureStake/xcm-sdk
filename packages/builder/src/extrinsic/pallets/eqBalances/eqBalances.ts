/* eslint-disable sort-keys */
import { ExtrinsicConfig } from '../../ExtrinsicConfig';
import { ExtrinsicConfigBuilder } from '../../ExtrinsicConfigBuilder.interfaces';

export enum EqBalancesFee {
  SovereignAccWillPay = 'SovereignAccWillPay',
  TargetChainAccWillPay = 'TargetChainAccWillPay',
  ThisAccWillPay = 'ThisAccWillPay',
}

const pallet = 'eqBalances';

export function eqBalances() {
  return {
    xcmTransfer: (): ExtrinsicConfigBuilder => ({
      build: ({ address, amount, asset, destination }) =>
        new ExtrinsicConfig({
          pallet,
          func: 'xcmTransfer',
          getArgs: () => [
            asset,
            amount,
            {
              parents: 1,
              interior: {
                X2: [
                  {
                    Parachain: destination.parachainId,
                  },
                  {
                    AccountKey20: {
                      network: 'Any',
                      key: address,
                    },
                  },
                ],
              },
            },
            EqBalancesFee.ThisAccWillPay,
          ],
        }),
    }),
    transferXcm: (): ExtrinsicConfigBuilder => ({
      build: ({ address, amount, asset, destination, fee, feeAsset }) =>
        new ExtrinsicConfig({
          pallet,
          func: 'transferXcm',
          getArgs: () => {
            const amountWithoutFee = amount - fee > 0n ? amount - fee : 0n;

            return [
              [asset, asset === feeAsset ? amountWithoutFee : amount],
              [feeAsset, fee],
              {
                parents: 1,
                interior: {
                  X2: [
                    {
                      Parachain: destination.parachainId,
                    },
                    {
                      AccountKey20: {
                        network: 'Any',
                        key: address,
                      },
                    },
                  ],
                },
              },
            ];
          },
        }),
    }),
  };
}
