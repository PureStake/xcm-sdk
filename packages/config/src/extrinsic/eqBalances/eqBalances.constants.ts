export enum EqBalancesExtrinsic {
  XcmTransfer = 'xcmTransfer',
  TransferXcm = 'transferXcm',
}

export enum EqBalancesSuccessEvent {
  ExtrinsicSuccess = 'ExtrinsicSuccess',
}

export enum EqBalancesFee {
  SovereignAccWillPay = 'SovereignAccWillPay',
  TargetChainAccWillPay = 'TargetChainAccWillPay',
  ThisAccWillPay = 'ThisAccWillPay',
}
