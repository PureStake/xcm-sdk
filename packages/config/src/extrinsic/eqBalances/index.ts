export * from './eqBalances';
export * from './eqBalances.constants';
export * from './eqBalances.interfaces';
