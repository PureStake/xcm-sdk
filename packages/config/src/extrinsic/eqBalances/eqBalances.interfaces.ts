import { XcmExtrinsicGetParams } from '../common.interfaces';
import { ExtrinsicPallet } from '../extrinsic.constants';
import {
  EqBalancesExtrinsic,
  EqBalancesSuccessEvent,
} from './eqBalances.constants';

export interface EqBalancesPallet {
  pallet: ExtrinsicPallet.EqBalances;
  extrinsic: EqBalancesExtrinsic;
  successEvent: EqBalancesSuccessEvent;
  getParams: (params: XcmExtrinsicGetParams) => EqBalancesPalletParams;
}

export type EqBalancesPalletParams = [
  /**
   * asset
   */
  number,
  /**
   * amount
   */
  bigint,
  /**
   * destination
   */
  {
    V1: {
      parents: 1;
      interior: {
        X2: [
          {
            Parachain: number;
          },
          {
            AccountKey20: {
              network: 'Any';
              /**
               * account
               */
              key: string;
            };
          },
        ];
      };
    };
  },
  /**
   * weight
   */
  'SovereignAccWillPay',
];
