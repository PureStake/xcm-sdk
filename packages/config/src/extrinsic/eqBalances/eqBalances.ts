import { MoonChain } from '../../interfaces';
import { ExtrinsicPallet } from '../extrinsic.constants';
import {
  EqBalancesExtrinsic,
  EqBalancesSuccessEvent,
} from './eqBalances.constants';
import { EqBalancesPallet } from './eqBalances.interfaces';

/* eslint-disable @typescript-eslint/no-use-before-define */
export function eqBalances(chain: MoonChain) {
  return {
    xcmTransfer: () => xcmTransfer(chain),
  };
}

function xcmTransfer(chain: MoonChain) {
  return {
    successEvent: (event: EqBalancesSuccessEvent) => ({
      asset: (id: number): EqBalancesPallet => ({
        pallet: ExtrinsicPallet.EqBalances,
        extrinsic: EqBalancesExtrinsic.XcmTransfer,
        successEvent: event,
        getParams: ({ account, amount }) => [
          id,
          amount,
          {
            parents: 1,
            interior: {
              X2: [
                {
                  Parachain: chain.parachainId,
                },
                {
                  AccountKey20: {
                    network: 'Any',
                    key: account,
                  },
                },
              ],
            },
          },
          'SovereignAccWillPay',
        ],
      }),
    }),
  };
}
