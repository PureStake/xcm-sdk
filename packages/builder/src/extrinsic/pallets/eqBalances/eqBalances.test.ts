import { buildParamsMock } from '../../../../fixtures';
import { eqBalances } from './eqBalances';

describe('eqBalances', () => {
  describe('xcmTransfer', () => {
    const extrinsic = eqBalances().xcmTransfer().build(buildParamsMock);

    it('should be correct config', () => {
      expect(extrinsic).toMatchSnapshot();
    });

    it('should get correct arguments', () => {
      expect(extrinsic.getArgs({} as any)).toMatchSnapshot();
    });
  });

  describe('transferXcm', () => {
    const extrinsic = eqBalances().transferXcm().build(buildParamsMock);

    it('should be correct config', () => {
      expect(extrinsic).toMatchSnapshot();
    });

    it('should get correct arguments', () => {
      expect(extrinsic.getArgs({} as any)).toMatchSnapshot();
    });
  });
});
