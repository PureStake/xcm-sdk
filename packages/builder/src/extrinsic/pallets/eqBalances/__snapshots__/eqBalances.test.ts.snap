// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`eqBalances transferXcm should be correct config 1`] = `
ExtrinsicConfig {
  "func": "transferXcm",
  "getArgs": [Function],
  "module": "eqBalances",
  "type": "Substrate",
}
`;

exports[`eqBalances transferXcm should get correct arguments 1`] = `
[
  [
    "USDT",
    99000000000n,
  ],
  [
    "RMRK",
    5000000000n,
  ],
  {
    "interior": {
      "X2": [
        {
          "Parachain": 1000,
        },
        {
          "AccountKey20": {
            "key": "0xeF46c7649270C912704fB09B75097f6E32208b85",
            "network": null,
          },
        },
      ],
    },
    "parents": 1,
  },
]
`;

exports[`eqBalances xcmTransfer should be correct config 1`] = `
ExtrinsicConfig {
  "func": "xcmTransfer",
  "getArgs": [Function],
  "module": "eqBalances",
  "type": "Substrate",
}
`;

exports[`eqBalances xcmTransfer should get correct arguments 1`] = `
[
  "USDT",
  99000000000n,
  {
    "interior": {
      "X2": [
        {
          "Parachain": 1000,
        },
        {
          "AccountKey20": {
            "key": "0xeF46c7649270C912704fB09B75097f6E32208b85",
            "network": null,
          },
        },
      ],
    },
    "parents": 1,
  },
  "ThisAccWillPay",
]
`;
